crates/wasm-encoder/src/component.rs

```diff
@@ -0,0 +1,371 @@
+    fn id(&self) -> u8;
+
+    /// Write this section's data and data length prefix into the given sink.
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>;
+}
+
+impl ComponentSection for RawSection<'_> {
+    fn id(&self) -> u8 {
+        self.id
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
+    {
+        <Self as Section>::encode(self, sink);
+    }
+}
+
+/// A Wasm adapter module that is being encoded.
+#[derive(Clone, Debug)]
+pub struct AdapterModule {
+    bytes: Vec<u8>,
+}
+
+impl AdapterModule {
+    /// Begin writing a new `AdapterModule`.
+    pub fn new() -> Self {
+        Self {
+            bytes: vec![
+                0x00, 0x61, 0x73, 0x6D, // magic (`\0asm`)
+                0x0a, 0x00, 0x01, 0x00, // version
+            ],
+        }
+    }
+
+    /// Finish writing this adapter module and extract ownership of the encoded bytes.
+    pub fn finish(self) -> Vec<u8> {
+        self.bytes
+    }
+
+    /// Write a section into this adapter module.
+    pub fn section(&mut self, section: &impl ComponentSection) -> &mut Self {
+        self.bytes.push(section.id());
+        section.encode(&mut self.bytes);
+        self
+    }
+}
+
+impl Default for AdapterModule {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A Wasm component that is being encoded.
+#[derive(Clone, Debug)]
+pub struct Component {
+    bytes: Vec<u8>,
+}
+
+impl Component {
+    /// Begin writing a new `Component`.
+    pub fn new() -> Self {
+        Self {
+            bytes: vec![
+                0x00, 0x61, 0x73, 0x6D, // magic (`\0asm`)
+                0x0a, 0x00, 0x02, 0x00, // version
+            ],
+        }
+    }
+
+    /// Finish writing this component and extract ownership of the encoded bytes.
+    pub fn finish(self) -> Vec<u8> {
+        self.bytes
+    }
+
+    /// Write a section into this component.
+    pub fn section(&mut self, section: &impl ComponentSection) -> &mut Self {
+        self.bytes.push(section.id());
+        section.encode(&mut self.bytes);
+        self
+    }
+}
+
+impl Default for Component {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Known component section IDs.
+///
+/// Useful for implementing the `ComponentSection` trait, or for setting
+/// `RawSection::id`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(u8)]
+pub enum SectionId {
+    /// The section is a custom section.
+    Custom = 0,
+    /// The section is a type section.
+    Type = 1,
+    /// The section is an import section.
+    Import = 2,
+    /// The section is a module section.
+    Module = 3,
+    /// The section is an instance section.
+    Instance = 4,
+    /// The section is an alias section.
+    Alias = 5,
+    /// The section is an export section.
+    Export = 6,
+    /// THe section is a function section.
+    Function = 7,
+    /// The section is an adapter function section.
+    AdapterFunction = 8,
```

Comment:
I would sort of naively expect that these would be part of the same section, but I also haven't caught up on the encoding in awhile.

Reply:
I think it makes sense to encode as two different sections as their expected types (IT-based vs. core based) as well as the "bodies" differ (currently in components, the bodies are simply pointing at what function is the target of the canonical lifting/lowering).

Reply:
Fixed the typo in the doc string on `Function` too.

Reply:
IIRC historically I concluded that having one notion of "function" was the way to go for interface types, as opposed to having two categories of functions as you're alluding to I think. I only vaguely remember now but I think it had to do with how to all embedders these things are just functions with fancier types, and it gets sort of redundant and repetitive to deal with two kinds of functions otherwise.

Also I was under the impression that components would want to manufacture core-wasm functions as adaptations of imported interface-types-using-functions to serve as imports for core-wasm imports?

Reply:
Perhaps we should discuss this more with Luke re: the format outlined [here](https://github.com/WebAssembly/interface-types/blob/component-rebase/design/proposals/interface-types/Binary.md)?